File: CRM-Client/src/schemas/opportunity.ts
/**
 * Zod Schema - Opportunity Types
 *
 * @description 商机类型 Schema，映射后端 schemas/opportunity.py
 */

import { z } from 'zod'
import { UserInfoSchema, PaginatedResponseSchema } from './common'

const NullableStringSchema = z.string().nullable()
const OptionalNullableStringSchema = z.string().nullable().optional()

export const OpportunityStatusSchema = z.number().int().min(0).max(2)
export const OpportunityApprovalPhaseSchema = z.enum(['draft', 'pending_review', 'approved', 'rejected'])
export const OpportunityPurchaseTypeSchema = z.enum(['NEW', 'RENEWAL', 'EXPANSION'])
export const OpportunityLicenseTypeSchema = z.enum(['SUBSCRIPTION', 'PERPETUAL'])

export const OpportunityOwnerInfoSchema = z.object({
  id: z.string(),
  name: z.string(),
  avatar_url: OptionalNullableStringSchema
}).passthrough()

export const OpportunityStageTemplateSchema = z.object({
  id: z.number().int(),
  stage_code: z.string(),
  stage_name: z.string(),
  win_probability: z.number(),
  sort_order: z.number(),
  is_default_start: z.boolean().optional(),
  can_skip: z.boolean().optional()
}).passthrough()

export const OpportunityProcurementMethodInfoSchema = z.object({
  id: z.number().int(),
  code: z.string().optional(),
  name: z.string(),
  description: OptionalNullableStringSchema,
  is_active: z.number().int().optional(),
  stages: z.array(OpportunityStageTemplateSchema).optional()
}).passthrough()

export const OpportunityCurrentStageSnapshotSchema = z.object({
  id: z.number().int(),
  stage_name: z.string(),
  win_probability: z.number(),
  template_sort_order: z.number(),
  template_code: z.string().nullable().optional(),
  entered_at: z.string(),
  exited_at: NullableStringSchema,
  procurement_method: z.object({
    id: z.number().int(),
    code: z.string(),
    name: z.string()
  }).passthrough().nullable().optional()
}).passthrough()

export const OpportunityApiResponseSchema = z.object({
  id: z.number().int(),
  opportunity_name: z.string(),
  customer_id: z.number().int(),
  customer_name: z.string().optional(),
  procurement_method_id: z.number().int().nullable(),
  procurement_method_info: OpportunityProcurementMethodInfoSchema.nullable().optional(),
  total_amount: z.number(),
  user_count: z.number().int(),
  unit_price: z.number(),
  license_type: OpportunityLicenseTypeSchema,
  subscription_years: z.number().nullable(),
  purchase_type: OpportunityPurchaseTypeSchema,
  decision_maker_count: z.number().int().nullable(),
  expected_closing_date: z.string(),
  procurement_stage_id: z.number().int().nullable(),
  stage_name: z.string().optional(),
  win_probability: z.number(),
  current_stage_snapshot: OpportunityCurrentStageSnapshotSchema.nullable().optional(),
  owner_id: z.string(),
  owner_info: OpportunityOwnerInfoSchema.optional(),
  creator_id: z.string(),
  creator_info: OpportunityOwnerInfoSchema.optional(),
  status: OpportunityStatusSchema,
  approval_phase: OpportunityApprovalPhaseSchema,
  actual_amount: z.number().nullable().optional(),
  actual_closing_date: z.string().nullable().optional(),
  loss_reason: z.string().nullable().optional(),
  created_time: z.string(),
  updated_time: z.string(),
  version: z.number().int(),
  customer_info: z.object({
    id: z.number().int(),
    account_name: z.string()
  }).passthrough().optional()
}).passthrough()

export const OpportunityListItemApiSchema = z.object({
  id: z.number().int(),
  opportunity_name: z.string(),
  customer_id: z.number().int(),
  procurement_method_id: z.number().int().nullable(),
  procurement_method_info: OpportunityProcurementMethodInfoSchema.nullable(),
  total_amount: z.number(),
  user_count: z.number().int(),
  unit_price: z.number(),
  license_type: z.string(),
  subscription_years: z.number().nullable(),
  purchase_type: z.string(),
  decision_maker_count: z.number().int().nullable(),
  expected_closing_date: z.string(),
  stage_id: z.number().int().nullable(),
  stage_name: z.string().optional(),
  win_probability: z.number(),
  owner_id: z.string(),
  creator_id: z.string(),
  current_stage_snapshot: OpportunityCurrentStageSnapshotSchema.nullable().optional(),
  stage: z.object({
    id: z.number().int(),
    stage_code: z.string(),
    stage_name: z.string(),
    win_probability: z.number(),
    sort_order: z.number(),
    description: NullableStringSchema,
    is_active: z.number().int(),
    created_time: z.string(),
    last_modified_time: z.string()
  }).passthrough().nullable(),
  stage_info: z.object({
    id: z.number().int(),
    stage_name: z.string(),
    win_probability: z.number(),
    is_default: z.number().int()
  }).passthrough().nullable(),
  status: OpportunityStatusSchema,
  approval_phase: OpportunityApprovalPhaseSchema,
  created_time: z.string(),
  last_modified_time: z.string(),
  customer_name: z.string().optional(),
  owner_info: OpportunityOwnerInfoSchema.optional()
}).passthrough()

export const OpportunityPaginatedListApiSchema = z.object({
  items: z.array(OpportunityListItemApiSchema),
  total: z.number().int().nonnegative(),
  page: z.number().int().positive(),
  page_size: z.number().int().positive(),
  total_pages: z.number().int().nonnegative()
}).passthrough()

export const OpportunityListApiResponseSchema = z.union([
  z.array(OpportunityListItemApiSchema),
  OpportunityPaginatedListApiSchema
])

export const SalesFunnelDataSchema = z.object({
  stage_id: z.number().int(),
  stage_name: z.string(),
  opportunity_count: z.number().int(),
  total_amount: z.number(),
  average_amount: z.number(),
  win_probability: z.number()
}).passthrough()

export const StageDurationDataSchema = z.object({
  stage_id: z.number().int(),
  stage_name: z.string(),
  avg_duration_days: z.number(),
  min_duration_days: z.number(),
  max_duration_days: z.number(),
  opportunity_count: z.number().int()
}).passthrough()

export const OpportunityOwnerFilterOptionsResponseSchema = z.object({
  data: z.array(z.object({
    id: z.string(),
    name: z.string(),
    is_me: z.boolean()
  }).passthrough())
}).passthrough()

// ===== 商机阶段枚举 =====
export const OpportunityStageSchema = z.enum([
  'INITIAL_CONTACT',    // 初步接洽
  'NEED_CONFIRMATION',  // 需求确认
  'SOLUTION_PROPOSAL',  // 方案报价
  'NEGOTIATION',        // 商务谈判
  'CONTRACT_SIGNING',   // 签约阶段
  'WON',                // 赢单
  'LOST'                // 输单
])

export const OpportunityStageMap: Record<string, string> = {
  'INITIAL_CONTACT': '初步接洽',
  'NEED_CONFIRMATION': '需求确认',
  'SOLUTION_PROPOSAL': '方案报价',
  'NEGOTIATION': '商务谈判',
  'CONTRACT_SIGNING': '签约阶段',
  'WON': '赢单',
  'LOST': '输单'
}

// ===== 商机基础类型 =====
export const OpportunityResponseSchema = z.object({
  id: z.number().int().positive(),
  opportunity_name: z.string().min(1).max(255),
  customer_id: z.number().int().positive(),
  expected_amount: z.number().positive().nullable(),
  expected_close_date: z.string().datetime().nullable(),
  stage: OpportunityStageSchema,
  probability: z.number().min(0).max(100).nullable(),
  owner_id: z.string().nullable(),
  source_lead_id: z.number().int().nullable(),
  lost_reason: z.string().nullable(),
  creator_id: z.string(),
  created_time: z.string().datetime(),
  last_modified_time: z.string().datetime(),
  version: z.number().int().nonnegative()
})

export type OpportunityResponse = z.infer<typeof OpportunityResponseSchema>

// ===== 商机列表响应 =====
export const OpportunityListResponseSchema = PaginatedResponseSchema(
  OpportunityResponseSchema.extend({
    owner_info: UserInfoSchema.nullable(),
    customer_name: z.string()
  })
)

export type OpportunityListResponse = z.infer<typeof OpportunityListResponseSchema>

// ===== 商机创建请求 =====
export const OpportunityCreateSchema = z.object({
  opportunity_name: z.string().min(1).max(255).optional(),
  customer_id: z.number().int().positive(),
  expected_amount: z.number().positive().optional(),
  expected_close_date: z.string().datetime().optional(),
  stage: OpportunityStageSchema.optional().default('INITIAL_CONTACT'),
  probability: z.number().min(0).max(100).optional(),
  source_lead_id: z.number().int().optional()
})

export type OpportunityCreate = z.infer<typeof OpportunityCreateSchema>

// ===== 商机更新请求 =====
export const OpportunityUpdateSchema = z.object({
  opportunity_name: z.string().min(1).max(255).optional(),
  expected_amount: z.number().positive().optional(),
  expected_close_date: z.string().datetime().optional(),
  stage: OpportunityStageSchema.optional(),
  probability: z.number().min(0).max(100).optional(),
  lost_reason: z.string().optional()
})

export type OpportunityUpdate = z.infer<typeof OpportunityUpdateSchema>

File: CRM-Server/app/schemas/opportunity.py
from pydantic import BaseModel, Field, field_validator, ConfigDict, model_validator
from typing import Optional, List
from datetime import date, datetime
from enum import Enum
from app.schemas.customer import ProcurementMethodInfo


class CurrentStageSnapshotInfo(BaseModel):
    """当前阶段快照信息"""
    id: int = Field(..., description="快照ID")
    procurement_stage_template_id: int = Field(..., description="阶段模板ID（与采购阶段列表的id对应）")
    stage_name: str = Field(..., description="阶段名称")
    win_probability: int = Field(..., description="阶段赢率（0-100）")
    template_sort_order: int = Field(..., description="阶段顺序（记录进入时模板的sort_order）")
    template_code: Optional[str] = Field(None, description="阶段模板编码")
    entered_at: datetime = Field(..., description="进入时间")
    exited_at: Optional[datetime] = Field(None, description="退出时间")
    procurement_method: Optional[ProcurementMethodInfo] = Field(None, description="采购方式信息")


class PurchaseTypeEnum(str, Enum):
    NEW = "NEW"
    RENEWAL = "RENEWAL"
    EXPANSION = "EXPANSION"
    
    @property
    def description(self):
        descriptions = {
            "NEW": "新购",
            "RENEWAL": "续购",
            "EXPANSION": "增购"
        }
        return descriptions.get(self.value, self.value)


class LicenseTypeEnum(str, Enum):
    SUBSCRIPTION = "SUBSCRIPTION"
    PERPETUAL = "PERPETUAL"
    
    @property
    def description(self):
        descriptions = {
            "SUBSCRIPTION": "订阅制",
            "PERPETUAL": "买断制"
        }
        return descriptions.get(self.value, self.value)


class OpportunityStatusEnum(str, Enum):
    FOLLOWING = "0"
    WON = "1"
    LOST = "2"
    
    @property
    def description(self):
        descriptions = {
            "0": "跟进中",
            "1": "赢单",
            "2": "输单"
        }
        return descriptions.get(self.value, self.value)


class OpportunityStageBase(BaseModel):
    stage_code: str = Field(..., min_length=1, max_length=100, description="阶段编码")
    stage_name: str = Field(..., min_length=1, max_length=100, description="阶段名称")
    win_probability: int = Field(..., ge=0, le=100, description="赢率（0-100）")
    sort_order: int = Field(..., ge=0, description="排序序号")
    description: Optional[str] = Field(None, max_length=500, description="阶段描述")


class OpportunityStageCreate(OpportunityStageBase):
    pass


class OpportunityStageUpdate(BaseModel):
    stage_name: Optional[str] = Field(None, min_length=1, max_length=100, description="阶段名称")
    win_probability: Optional[int] = Field(None, ge=0, le=100, description="赢率（0-100）")
    sort_order: Optional[int] = Field(None, ge=0, description="排序序号")
    description: Optional[str] = Field(None, max_length=500, description="阶段描述")
    is_active: Optional[int] = Field(None, ge=0, le=1, description="是否启用")


class OpportunityStageResponse(BaseModel):
    id: int = Field(..., description="销售阶段ID")
    stage_code: str = Field(..., description="阶段编码")
    stage_name: str = Field(..., description="阶段名称")
    win_probability: int = Field(..., description="赢率（0-100）")
    sort_order: int = Field(..., description="排序序号")
    description: Optional[str] = Field(None, description="阶段描述")
    is_active: int = Field(..., description="是否启用：0:否, 1:是")
    created_time: datetime = Field(..., description="创建时间")
    last_modified_time: datetime = Field(..., description="最后修改时间")
    
    model_config = ConfigDict(from_attributes=True)


class OpportunityBase(BaseModel):
    opportunity_name: str = Field(..., min_length=1, max_length=255, description="商机名称（项目名称）")
    total_amount: float = Field(..., gt=0, description="预计总金额（元）")
    user_count: int = Field(..., gt=0, description="采购用户数（ licenses 数量）")
    license_type: LicenseTypeEnum = Field(..., description="授权模式：SUBSCRIPTION:订阅制, PERPETUAL:买断制")
    subscription_years: Optional[int] = Field(None, gt=0, description="订阅年限（订阅制时必填，默认1年）")
    purchase_type: PurchaseTypeEnum = Field(..., description="采购类型：NEW:新购, RENEWAL:续购, EXPANSION:增购")
    decision_maker_count: Optional[int] = Field(None, ge=1, description="采购决策人数（影响销售策略）")
    expected_closing_date: date = Field(..., description="预计成交日期")
    
    @model_validator(mode='after')
    def validate_license_fields(cls, values):
        license_type = values.license_type
        subscription_years = values.subscription_years
        
        if license_type == LicenseTypeEnum.SUBSCRIPTION:
            if subscription_years is None or subscription_years <= 0:
                raise ValueError('订阅制下订阅年限必须大于0')
        
        return values
    
    @field_validator('opportunity_name')
    @classmethod
    def opportunity_name_must_not_be_empty(cls, v):
        if v is not None and (not v or not v.strip()):
            raise ValueError('商机名称不能为空')
        return v.strip() if v else v


class OpportunityCreate(OpportunityBase):
    customer_id: int = Field(..., description="关联客户ID（关联到 crm_customers 表）")
    procurement_method_id: Optional[int] = Field(None, description="采购方式ID（关联到 crm_procurement_methods 表），如果不指定则使用客户的默认采购方式")
    stage_id: Optional[int] = Field(None, description="初始销售阶段ID（关联到 crm_procurement_stage_templates 表），如果不指定则使用采购方式的默认起始阶段")
    owner_id: Optional[str] = Field(None, description="负责人系统用户ID，如果不指定则默认为创建人")


class OpportunityUpdate(BaseModel):
    opportunity_name: Optional[str] = Field(None, min_length=1, max_length=255, description="商机名称（项目名称）")
    total_amount: Optional[float] = Field(None, gt=0, description="预计总金额（元）")
    user_count: Optional[int] = Field(None, gt=0, description="采购用户数（ licenses 数量）")
    license_type: Optional[LicenseTypeEnum] = Field(None, description="授权模式：SUBSCRIPTION:订阅制, PERPETUAL:买断制")
    subscription_years: Optional[int] = Field(None, gt=0, description="订阅年限（订阅制时必填，默认1年）")
    purchase_type: Optional[PurchaseTypeEnum] = Field(None, description="采购类型：NEW:新购, RENEWAL:续购, EXPANSION:增购")
    decision_maker_count: Optional[int] = Field(None, ge=1, description="采购决策人数（影响销售策略）")
    expected_closing_date: Optional[date] = Field(None, description="预计成交日期")
    procurement_method_id: Optional[int] = Field(None, description="采购方式ID（关联到 crm_procurement_methods 表）")
    owner_id: Optional[str] = Field(None, description="负责人系统用户ID")
    
    @model_validator(mode='after')
    def validate_license_fields(cls, values):
        license_type = values.license_type
        subscription_years = values.subscription_years
        
        if license_type == LicenseTypeEnum.SUBSCRIPTION:
            if subscription_years is None or subscription_years <= 0:
                raise ValueError('订阅制下订阅年限必须大于0')
        
        return values


class OpportunityStageUpdate(BaseModel):
    stage_template_id: int = Field(..., description="目标采购阶段模板ID（推进到下一阶段）")


class OpportunityMoveToStage(BaseModel):
    stage_template_id: int = Field(..., description="目标采购阶段模板ID（推进到下一阶段）")


class OpportunityWin(BaseModel):
    actual_amount: float = Field(..., gt=0, description="实际成交金额（元）")
    actual_closing_date: date = Field(..., description="实际成交日期")


class OpportunityLose(BaseModel):
    loss_reason: str = Field(..., min_length=1, max_length=500, description="输单原因（必填，用于后续分析）")
    
    @field_validator('loss_reason')
    @classmethod
    def loss_reason_must_not_be_empty(cls, v):
        if v is not None and (not v or not v.strip()):
            raise ValueError('输单原因不能为空')
        return v.strip() if v else v


class OpportunityResponse(BaseModel):
    id: int = Field(..., description="商机ID（主键）")
    opportunity_name: str = Field(..., description="商机名称（项目名称）")
    customer_id: int = Field(..., description="关联客户ID")
    procurement_method_id: Optional[int] = Field(None, description="采购方式ID")
    procurement_method_info: Optional[ProcurementMethodInfo] = Field(None, description="采购方式详细信息")
    total_amount: float = Field(..., description="预计总金额（元）")
    user_count: int = Field(..., description="采购用户数（ licenses 数量）")
    unit_price: float = Field(..., description="标准单价（系统自动计算：订阅制=总金额/用户数/年数，买断制=总金额/用户数/5）")
    license_type: str = Field(..., description="授权模式：SUBSCRIPTION:订阅, PERPETUAL:买断")
    subscription_years: Optional[int] = Field(None, description="订阅年限（订阅制时必填）")
    purchase_type: str = Field(..., description="采购类型：NEW:新购, RENEWAL:续购, EXPANSION:增购")
    decision_maker_count: Optional[int] = Field(None, description="采购决策人数（影响销售策略）")
    expected_closing_date: date = Field(..., description="预计成交日期")
    stage_id: Optional[int] = Field(None, description="当前销售阶段ID（已废弃，保留用于兼容）")
    win_probability: int = Field(..., description="当前阶段赢率（0-100，继承自销售阶段配置）")
    owner_id: str = Field(..., description="负责人系统用户ID")
    status: int = Field(..., description="商机状态：0:跟进中, 1:已赢单, 2:已输单")
    loss_reason: Optional[str] = Field(None, description="输单原因（status=2时有值）")
    actual_amount: Optional[float] = Field(None, description="实际成交金额（status=1时有值）")
    actual_closing_date: Optional[date] = Field(None, description="实际成交日期（status=1时有值）")
    creator_id: str = Field(..., description="创建人系统用户ID")
    created_time: datetime = Field(..., description="创建时间")
    last_modified_time: datetime = Field(..., description="最后修改时间")
    version: int = Field(..., description="版本号（乐观锁，防止并发修改冲突）")
    current_stage_snapshot: Optional[CurrentStageSnapshotInfo] = Field(None, description="当前阶段快照信息")
    
    model_config = ConfigDict(from_attributes=True)


class CustomerInfo(BaseModel):
    id: int = Field(..., description="客户ID")
    account_name: str = Field(..., description="客户公司名称")
    industry: Optional[str] = Field(None, description="所属行业")
    city: str = Field(..., description="所在城市")
    address: Optional[str] = Field(None, description="公司地址")
    company_scale: Optional[str] = Field(None, description="公司规模")
    status: int = Field(..., description="客户状态")
    owner_id: Optional[str] = Field(None, description="负责人系统用户ID")
    
    model_config = ConfigDict(from_attributes=True)


class OwnerInfo(BaseModel):
    id: str = Field(..., description="系统用户ID")
    name: str = Field(..., description="用户姓名")
    avatar_url: Optional[str] = Field(None, description="用户头像URL")
    
    model_config = ConfigDict(from_attributes=True)


class OpportunityProcurementStageInfo(BaseModel):
    """商机采购阶段信息"""
    id: int = Field(..., description="阶段模板ID")
    stage_name: str = Field(..., description="阶段名称")
    win_probability: int = Field(..., description="阶段赢率（0-100）")
    sort_order: int = Field(..., description="阶段顺序")
    is_current: bool = Field(..., description="是否为当前商机的阶段")
    is_default_start: bool = Field(..., description="是否为默认起始阶段")
    can_skip: bool = Field(..., description="是否可跳过")


class OpportunityDetailResponse(OpportunityResponse):
    current_stage_snapshot: Optional[CurrentStageSnapshotInfo] = Field(None, description="当前阶段快照信息")
    procurement_stages: Optional[List[OpportunityProcurementStageInfo]] = Field(None, description="该采购方式的所有阶段列表")
    customer_name: Optional[str] = Field(None, description="客户名称")
    customer_info: Optional['CustomerInfo'] = Field(None, description="客户详细信息")
    owner_info: Optional[OwnerInfo] = Field(None, description="负责人信息")
    creator_info: Optional[OwnerInfo] = Field(None, description="创建人信息")


class OpportunityListResponse(OpportunityResponse):
    stage: Optional[OpportunityStageResponse] = Field(None, description="销售阶段信息")
    customer_name: Optional[str] = Field(None, description="客户名称")
    owner_info: Optional[OwnerInfo] = Field(None, description="负责人信息")


class MessageResponse(BaseModel):
    message: str = Field(..., description="响应消息")


class SalesFunnelResponse(BaseModel):
    stage_id: int = Field(..., description="销售阶段ID")
    stage_code: str = Field(..., description="阶段编码")
    stage_name: str = Field(..., description="阶段名称")
    win_probability: int = Field(..., description="赢率（0-100）")
    opportunity_count: int = Field(..., description="商机数量")
    total_amount: float = Field(..., description="总金额")
    average_amount: float = Field(..., description="平均金额")


class StageDurationResponse(BaseModel):
    stage_id: int = Field(..., description="销售阶段ID")
    stage_code: str = Field(..., description="阶段编码")
    stage_name: str = Field(..., description="阶段名称")
    average_days: float = Field(..., description="平均停留天数")
    min_days: float = Field(..., description="最短停留天数")
    max_days: float = Field(..., description="最长停留天数")
    opportunity_count: int = Field(..., description="商机数量")


class AnalyticsFilter(BaseModel):
    owner_id: Optional[str] = Field(None, description="负责人ID")
    start_date: Optional[date] = Field(None, description="开始日期")
    end_date: Optional[date] = Field(None, description="结束日期")
